### ISSUE TITLE
helper function for selecting a gpu in multi-gpu setting

### ISSUE THREAD

**Issue by CarloLucibello** [FluxML/Flux.jl#1074]
As per discussion in https://github.com/JuliaGPU/CuArrays.jl/issues/52, that I report here, the current way to use a specific gpu in a Flux script would be the following
```julia
gpu_id = 0  ## set < 0 for no cuda, >= 0 for using a specific device (if available)

if CUDAapi.has_cuda_gpu() && gpu_id >=0
    CUDAdrv.device!(gpu_id)
    CuArrays.allowscalar(false)
    device = Flux.gpu
    @info "Training on GPU-$(gpu_id)"
else
    device = Flux.cpu
    @info "Training on CPU"
end

model = model |> device
for x in data
    x = x |> device
    ....
```
This involves importing 3 cuda libraries,  CUDAapi, CuArrays and CUDAdrv, which likely won't be likely used in the rest of the script. 

I suggest to wrap the same functionality under Flux, in a  `gpu! ` function, so that we can equivalently write
```julia
gpu_id = 0  ## set < 0 for no cuda, >= 0 for using a specific device (if available)

if CUDAapi.has_cuda_gpu() && gpu_id >=0
    device = Flux.gpu!(gpu_id, allowscalar=false)
    @info "Training on GPU-$(gpu_id)"
else
    device = Flux.cpu
    @info "Training on CPU"
end
```
Thoughts?

**Comment by dhairyagandhi96** [comment]
```julia
CuArrays.device!(id) do
  ... # the usual
end
```
Should be equivalent and cleaner

**Comment by CarloLucibello** [comment]
you mean wrapping model creation and training within the do block?

**Comment by dhairyagandhi96** [comment]
Yeah, most times you want to be data parallel, so having multiple workers start tasks on different gpus and syncing appropriately is usually preferred.

**Comment by CarloLucibello** [comment]
> See also: FluxML/Hydra.jl

what should I look at?

**Comment by dhairyagandhi96** [comment]
It's way to move model training keepings things in sync has been useful in the past to avoid waiting around too long when you have heterogenous devices in the same machine

**Comment by CarloLucibello** [comment]
multi-gpu training and data parallelism is something we definitely want, but this issue is about the much simpler problem of selecting a gpu for single gpu training in multi-gpu setting. 
For that, it just seems convenient to have `Flux.gpu!(gpu_id)`.

**Comment by johnnychen94** [comment]
```julia
Flux.device!(CUDA(gpu_id))
Flux.device!(CPU())
```

looks more intuitive to me

https://github.com/timholy/ComputationalResources.jl/blob/master/src/ComputationalResources.jl#L127-L129
